src/util/postgres_client.js

```diff
@@ -1207,27 +1207,29 @@ class PostgresTable {
         // eslint-disable-next-line no-constant-condition
         while (true) {
             let pg_client;
+            let locked;
             try {
                 pg_client = await this.client.pool.connect();
-                let update_res = await this._updateOneWithClient(pg_client, query, update, options);
-                if (update_res.rowCount === 0) {
-                    // try to lock the advisory_lock_key for this table and insert the first doc
-                    let locked = await this.try_lock_table(pg_client);
-                    if (locked) {
+                locked = await this.try_lock_table(pg_client);
+                if (locked) {
```

Comment:
the purpose of the update before the lock is to avoid serialization of all upserts. I think that the bug here is that we need to try and update again after the lock is acquired and before inserting